### PR TITLE
ztp: OCPBUGS-3389: SiteConfig does not support modifying install conf…

### PR DESCRIPTION
**Pull request by sabbir-47**
Introduces a new field named installConfigOverrides in the Siteconfig, which allows to pass the install config parameters to be set at annotations in AgentClusterInstall CR.Unit tests have been adapted to validate the changes

Signed-off-by: Sabbir Hasan <sahasan@redhat.com>

/cc @imiller0 @serngawy @browsell

### CODE REVIEW

File: ztp/siteconfig-generator/siteConfig/siteConfigBuilder.go [hunk 1]
```diff
@@ -75,12 +76,16 @@ func (scbuilder *SiteConfigBuilder) validateSiteConfig(siteConfigTemp SiteConfig
...
-		if cluster.NetworkType != "OpenShiftSDN" && cluster.NetworkType != "OVNKubernetes" {
-			return errors.New("Error: networkType must be either OpenShiftSDN or OVNKubernetes " + siteConfigTemp.Metadata.Name + "/" + cluster.ClusterName)
+		// will verify if installConfigOvverrides is a valid json if not empty
+		if cluster.InstallConfigOverrides != "" {
+			if !json.Valid([]byte(cluster.InstallConfigOverrides)) {
+				return errors.New("Error: Invalid json parameter set at installConfigOverride for " + siteConfigTemp.Metadata.Name + "/" + cluster.ClusterName)
+			}
+			siteConfigTemp.Spec.Clusters[id].NetworkType = cluster.InstallConfigOverrides
+		} else {
+			siteConfigTemp.Spec.Clusters[id].NetworkType = "{\"networking\":{\"networkType\":\"" + cluster.NetworkType + "\"}}"
```
Comment (imiller0) [inline ztp/siteconfig-generator/siteConfig/siteConfigBuilder.go:86]: A user could include both networkType and installConfigOverrides in the same SiteConfig. I think in that case the expected behavior would be to merge the JSON content. 
If the user enters this in the siteconfig:
```
installConfigOverrides: |
  {
    "features" : {
      "first" : "abc",
      "second" : "def"
    }
  }
  networkType: OVNKubernetes
```

The resulting JSON would be:
```
{
  "features" : {
    "first" : "abc",
    "second" : "def"
  },
  "networking" : {
    "networkType": "OVNKubernetes"
  }
}
```

Comment (sabbir-47) [reply ztp/siteconfig-generator/siteConfig/siteConfigBuilder.go]: @imiller0 fixed as suggested, added additional unit test.

File: ztp/siteconfig-generator/siteConfig/siteConfigHelper.go [hunk 2]
```diff
@@ -173,3 +174,28 @@ func deleteInspectAnnotation(bmhCR map[string]interface{}) map[string]interface{
...
+		return "", fmt.Errorf("Invalid json conversion of network type")
+	}
+
+	if configOverrides != "" {
+		if !json.Valid([]byte(configOverrides)) {
+			return "", fmt.Errorf("Invalid json parameter set at installConfigOverride")
+		}
+		trimmedConfigOverrides := strings.TrimPrefix(configOverrides, "{")
+		trimmedNetworkType := strings.TrimSuffix(networkAnnotation, "}")
+		finalJson := trimmedNetworkType + "," + trimmedConfigOverrides
```
Comment (imiller0) [inline ztp/siteconfig-generator/siteConfig/siteConfigHelper.go:192]: What will happen if the "network" is also included in the installConfigOverrides, perhaps with other non-overlapping networking settings?

File: ztp/ran-crd/site-config-crd.yaml [hunk 3]
```diff
@@ -112,6 +112,9 @@ spec:
                         enum:
                           - OVNKubernetes
                           - OpenShiftSDN
+                      installConfigOverrides:
+                        description: InstallConfigOverrides allows to pass install config parameters to be added as annotation in the AgentClusterInstall CR.
```
Comment (imiller0) [inline ztp/ran-crd/site-config-crd.yaml:116]: Maybe include the specific annotation in the description to help users understand where/what is affected.

File: ztp/siteconfig-generator/siteConfig/siteConfigHelper.go [hunk 4]
```diff
@@ -173,3 +174,28 @@ func deleteInspectAnnotation(bmhCR map[string]interface{}) map[string]interface{
 	}
 	return bmhCR
 }
+
+// buildJsonAnnotation returns string in json format
+func buildJsonAnnotation(networkType, configOverrides string) (string, error) {
```
[the remark this answers is not in the document]
Comment (sabbir-47) [reply ztp/siteconfig-generator/siteConfig/siteConfigHelper.go:179]: Will rename in the next fix. thanks

File: ztp/ran-crd/site-config-crd.yaml (hunk 3 above)
Comment (sabbir-47) [reply ztp/ran-crd/site-config-crd.yaml:116]: Will add in the description in the next fix.

File: ztp/siteconfig-generator/siteConfig/siteConfigHelper.go (hunk 2 above)
Comment (sabbir-47) [reply ztp/siteconfig-generator/siteConfig/siteConfigHelper.go:192]: For example, if we have below in siteconfig:
```
networkType: OVNKubernetes
installConfigOverrides "{\"networking\":{\"UserManagedNetworking\":\"True\"}}"`
```
Do we want the annotation to be for example in array format for the networking:

```
{"networking":[{"networkType":"OVNKubernetes"},{"UserManagedNetworking":"True"}]}
```

Comment (imiller0) [reply ztp/siteconfig-generator/siteConfig/siteConfigHelper.go:192]: The content of this JSON string (the entire overrides) feeds into the [installConfig defined here](https://github.com/openshift/installer/blob/53727e66b810286f5f10441470623f0a8bb11b1c/pkg/types/installconfig.go#L87) (note the Network on line 115). The Network isn't a list/array, so I don't think we can embed it as a list entry. If the user gives us a networkType I think you can take what is in your array and do a "dictionary merge" with any other networking content.

File: ztp/siteconfig-generator/siteConfig/siteConfigHelper.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (sabbir-47) [reply ztp/siteconfig-generator/siteConfig/siteConfigHelper.go]: Fixed with commit 0843a14efa1eb6c8e7d476a1d7197a790758d76f

File: ztp/siteconfig-generator/siteConfig/siteConfigHelper.go (hunk 2 above)
Comment (sabbir-47) [reply ztp/siteconfig-generator/siteConfig/siteConfigHelper.go]: Fixed with commit 0843a14efa1eb6c8e7d476a1d7197a790758d76f, please look at the tests that are added.

File: ztp/ran-crd/site-config-crd.yaml (hunk 3 above)
Comment (sabbir-47) [reply ztp/ran-crd/site-config-crd.yaml]: description added with commit 0843a14efa1eb6c8e7d476a1d7197a790758d76f

File: ztp/siteconfig-generator/siteConfig/siteConfigHelper_test.go [hunk 5]
```diff
@@ -162,3 +162,72 @@ func Test_addZTPAnnotationToManifest(t *testing.T) {
...
+			error:                 nil,
+			name:                  "installConfigOverride contains non-overlapping networking settings",
+		},
+
+		{
+			networkType:           "OVNKubernetes",
+			installConfigOverride: "{\"networking\":{\"UserManagedNetworking\":\"True\",\"networkType\":\"default\"},\"features\":[{\"abc\":\"test\"},{\"xyz\":\"test1\"}]}",
+			expected:              "{\"features\":[{\"abc\":\"test\"},{\"xyz\":\"test1\"}],\"networking\":{\"UserManagedNetworking\":\"True\",\"networkType\":\"OVNKubernetes\"}}",
+			error:                 nil,
+			name:                  "installConfigOverride contains bad networking settings",
```
Comment (imiller0) [inline ztp/siteconfig-generator/siteConfig/siteConfigHelper_test.go:219]: It would be good to have one test where the user does not include networkType at all (or is empty) for completeness.

**Review by imiller0 — COMMENTED** [review]
Can you add one more test? Other than that it looks good. Thanks!

File: ztp/siteconfig-generator/siteConfig/siteConfigBuilder_test.go [hunk 6]
```diff
@@ -545,6 +613,15 @@ spec:
 	}
 }
 
+func Test_SNOClusterSiteConfigBuildWithoutNetworkType(t *testing.T) {
```
Comment (sabbir-47) [inline ztp/siteconfig-generator/siteConfig/siteConfigBuilder_test.go:616]: @imiller0 I have added this test  where networkType wasn't set

File: ztp/siteconfig-generator/siteConfig/siteConfigHelper_test.go (hunk 5 above)
Comment (sabbir-47) [reply ztp/siteconfig-generator/siteConfig/siteConfigHelper_test.go:219]: If i add a test with networkType set to empty, that will fail because we can't mimic the CR behavior here. I have added a test in siteConfigBuilder_test.go where networkType is not set in the CR and it takes the default value which is defined in the CRD and return the desired behavior.

File: ztp/siteconfig-generator/siteConfig/siteConfigBuilder_test.go (hunk 6 above)
Comment (imiller0) [reply ztp/siteconfig-generator/siteConfig/siteConfigBuilder_test.go:616]: Ah, ok and siteConfigTestWithoutNetworkType includes an installConfigOverride. Perfect, thank you!

**Review by imiller0 — APPROVED** [review]
/lgtm
